extensions/poetry/main.ts

```diff
@@ -0,0 +1,98 @@
+import { PhylumApi } from "phylum"
+import { red, green, yellow } from "https://deno.land/std@0.150.0/fmt/colors.ts";
+
+class FileBackup {
+  readonly fileName: string
+  readonly fileContent: string | null
+
+  constructor(fileName: string) {
+    this.fileName = fileName
+    this.fileContent = null
+  }
+
+  async backup() {
+    try {
+      this.fileContent = await Deno.readTextFile(this.fileName)
+    } catch (e) { }
+  }
+
+  async restoreOrDelete() {
+    try {
+      if (this.fileContent != null) {
+        await Deno.writeTextFile(this.fileName, this.fileContent)
+      } else {
+        await Deno.remove(this.fileName)
+      }
+    } catch (e) { }
+  }
+}
+
+// Parse the output of `poetry` subcommands that support the `--dry-run` flag,
+// and submit the packages.
+//
+// Add the `-vvv` flags to get a detailed report of the dependency resolution
+// process. This way, we can track the actual lockfile changes, as specifying
+// `--dry-run` only would not output anything in combination with flags such
+// as `--lock` that do not perform the actual operations.
+async function poetryCheckDryRun(subcommand: string, args: string[]): boolean {
+  // Read and backup the current poetry lockfile contents.
+  const lockfileBackup = new FileBackup('poetry.lock')
+  const manifestBackup = new FileBackup('pyproject.toml')
+
+  await lockfileBackup.backup()
+  await manifestBackup.backup()
+
+  let process = Deno.run({
+    cmd: ['poetry', subcommand, '-vvv', '-n', '--dry-run', ...args.map(s => s.toString())],
+    stdout: 'piped',
+    stderr: 'piped',
+  })
+
+  await process.status()
+  await process.close()
+
+  const lockfileData = await PhylumApi.parseLockfile('./poetry.lock', 'poetry')
+
+  // If it existed before, restore the previous contents of the lockfile;
+  // otherwise, delete it. This is a workaround to the fact that in poetry
+  // 1.1.x, the `--dry-run` argument does not prevent the lockfile from 
+  // being modified. This is not fixed as of poetry 1.1.14.
+  // Prudently, do the same for the manifest (pyproject.toml).
+  //
+  // See: https://github.com/python-poetry/poetry/pull/5718
+  await lockfileBackup.restoreOrDelete()
+  await manifestBackup.restoreOrDelete()
+
+  const jobId = await PhylumApi.analyze(lockfileData['package_type'], lockfileData['packages'])
```

Comment:
Out of curiosity, why did you index `lockfileData` using the map-type brackets, rather than just `lockfileData.package_type`. Did the latter give you a warning or something?

Reply:
Oh that's actually a good point. Maybe we should change this? The type is constructed specifically for our extension API, so maybe it should be `packageType` and `packages` in Rust instead?

Reply:
If we can do it easily (e.g. `#[serde(rename = "packageType"]`), yeah, otherwise we are going to add camel case variable names and Clippy is not going to be happy about it. I _think_ Deno uses `serde` for v8 types so this should be feasible.